library/alloc/src/raw_rc.rs

```diff
@@ -0,0 +1,2468 @@
+//!   `size_of::<RefCounts>().max(align_of::<T>()) - align_of::<T>().saturating_sub(size_of::<RefCounts>())`,
+//!   which equals to `size_of::<RefCounts>()`, which is a fixed offset and ensures that the
+//!   `RefCounts` object and the value object does not overlap.
+//! - The size of the allocation is `size_of::<RefCounts>().max(align_of::<T>()) + size_of::<T>()`,
+//!   and the value is stored at `size_of::<RefCounts>().max(align_of::<T>())`, which leaves exactly
+//!   `size_of::<T>()` bytes for storing the value object.
+//!
+//! So both the `RefCounts` object and the value object have their alignment and size requirements.
+//! And we get a fixed offset between those two objects.
+//!
+//! # Reference-counting pointer design
+//!
+//! Both strong and weak reference-counting pointers stores a pointer that points to the value
+//! object in a reference counted allocation instead of the beginning of the allocation, this is
+//! base on the assumption that users access the contained value more frequently than the reference
+//! counters. Also, this possibly allows us to enable some future optimizations like:
+//!
+//! - Making reference-counting pointers have ABI-compatible representation as raw pointers so we
+//!   can use them directly in FFI interfaces.
+//! - Converting `Option<Rc<T>>` to `Option<&T>` with memory copy operation.
+//! - Converting `&[Rc<T>]` to `&[&T]` with zero cost.
+
+use core::alloc::{AllocError, Allocator, Layout, LayoutError};
+use core::any::Any;
+use core::cell::UnsafeCell;
+#[cfg(not(no_global_oom_handling))]
+use core::clone::CloneToUninit;
+use core::error::{Error, Request};
+use core::fmt::{self, Debug, Display, Formatter, Pointer};
+use core::hash::{Hash, Hasher};
+#[cfg(not(no_global_oom_handling))]
+use core::iter::TrustedLen;
+use core::marker::{PhantomData, Unsize};
+#[cfg(not(no_global_oom_handling))]
+use core::mem::ManuallyDrop;
+use core::mem::{self, MaybeUninit, SizedTypeProperties};
+use core::num::NonZeroUsize;
+use core::ops::{CoerceUnsized, Deref, DispatchFromDyn};
+use core::pin::PinCoerceUnsized;
+use core::ptr::{self, NonNull};
+use core::str;
+
+#[cfg(not(no_global_oom_handling))]
+use crate::alloc;
+use crate::alloc::Global;
+#[cfg(not(no_global_oom_handling))]
+use crate::boxed::Box;
+#[cfg(not(no_global_oom_handling))]
+use crate::string::String;
+#[cfg(not(no_global_oom_handling))]
+use crate::vec::Vec;
+
+/// The return value type for `RcOps::make_mut`.
+#[cfg(not(no_global_oom_handling))]
+pub(crate) enum MakeMutStrategy {
+    /// This `RawRc` is the only strong pointer that reference to the value, but there are weak
+    /// pointers also reference to the value. Before returning, the strong reference count has been
+    /// set to zero to prevent new strong pointers being created through upgrading from weak
+    /// pointers.
+    Move,
+    /// There is more than one strong pointers reference to the value.
+    Clone,
```

Comment:
```suggestion
    /// This `RawRc` is the only strong pointer that references the value, but there are weak
    /// pointers also referencing the value. Before returning, the strong reference count has been
    /// set to zero to prevent new strong pointers being created through upgrading from weak
    /// pointers.
    Move,
    /// There is more than one strong pointer that references the value.
    Clone,
```